rust-runtime/aws-smithy-http-server-python/src/tls/listener.rs

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::convert::Infallible;
+use std::pin::Pin;
+use std::sync::mpsc;
+use std::task::{Context, Poll};
+
+use futures::{ready, Stream};
+use hyper::server::accept::Accept;
+use pin_project_lite::pin_project;
+use tls_listener::{AsyncAccept, AsyncTls, TlsListener};
+
+pin_project! {
+    /// A wrapper around [TlsListener] that allows changing TLS config via a channel
+    /// and ignores incorrect connections (they cause Hyper server to shutdown otherwise).
+    pub struct Listener<A: AsyncAccept, T: AsyncTls<A::Connection>> {
+        #[pin]
+        inner: TlsListener<A, T>,
+        new_acceptor_rx: mpsc::Receiver<T>,
+    }
+}
+
+impl<A: AsyncAccept, T: AsyncTls<A::Connection>> Listener<A, T> {
+    pub fn new(tls: T, listener: A, new_acceptor_rx: mpsc::Receiver<T>) -> Self {
+        Self {
+            inner: TlsListener::new(tls, listener),
+            new_acceptor_rx,
+        }
+    }
+}
+
+impl<A, T> Accept for Listener<A, T>
+where
+    A: AsyncAccept,
+    A::Error: std::error::Error,
+    T: AsyncTls<A::Connection>,
+{
+    type Conn = T::Stream;
+    type Error = Infallible;
+
+    fn poll_accept(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Conn, Self::Error>>> {
+        // Replace current acceptor (it also contains TLS config) if there is a new one
+        if let Ok(acceptor) = self.new_acceptor_rx.try_recv() {
+            self.as_mut().project().inner.replace_acceptor_pin(acceptor);
+        }
+
+        loop {
+            match ready!(self.as_mut().project().inner.poll_next(cx)) {
+                Some(conn) => match conn {
+                    Ok(conn) => return Poll::Ready(Some(Ok(conn))),
+                    Err(err) => {
+                        // Don't propogate errors to Hyper because it causes server to shutdown
+                        tracing::error!(error = ?err, "tls connection error");
```

Comment:
This is going to blast us with `tracing::error` if the inner `TlsListener<A, T>` gets into a state where `poll_next` always returns `Some(Err(_))`. Is this possible? Do we want to add a delay or shutdown hyper?

Reply:
Good point, inner `TlsListener` returns either an `TlsAcceptor::Error` or a TLS handshake error: https://github.com/tmccombs/tls-listener/blob/main/src/lib.rs#L143-L152

I believe handshake errors shouldn't cause inner listener to stuck, so I've changed to propagate `TlsAcceptor::Error`s to upstream and reduced TLS handshake error logs to `DEBUG` level: https://github.com/awslabs/smithy-rs/pull/2002/commits/1710550576e369b18cd0a16015a04118d20b95ec

Reply:
I'm unsure of when a `ListenerError` vs a `TlsAcceptError` occurs. If clients can cause a `ListenerError` that's a problem I'm guessing.

Reply:
I'm also not sure but I'm looking to `warp` and seems like they also propagate listener errors: https://github.com/seanmonstar/warp/blob/4e9c4fd6ce238197fd1088061bbc07fa2852cb0f/src/tls.rs#L381

Reply:
Maybe it's worth just doing a little check like scrambling the TLS cert for client/server and see what happens.

Reply:
I've adde some tests with https://github.com/awslabs/smithy-rs/pull/2002/commits/ce6a85e9600eacb3802547f0ee7ef6e8e082f70a